### PR TITLE
[torchao]: fix how FqnConfig is resolved and handled

### LINKED ISSUE #14667: `FqnToConfig` raises inside `TorchAoConfig`
### Describe the bug

seems like it's not applied to quantize_ properly

### Reproduction

```py
from collections import OrderedDict

import torch
from diffusers.models.transformers.transformer_flux2 import Flux2Transformer2DModel
from diffusers.quantizers.quantization_config import TorchAoConfig
from torchao.quantization import FqnToConfig, Int8WeightOnlyConfig, PerGroup, PerTensor

with torch.no_grad():
    _t = Flux2Transformer2DModel.from_pretrained(
        "black-forest-labs/FLUX.2-klein-4B",
        subfolder="transformer",
        torch_dtype=torch.bfloat16,
        device_map="cuda",
        low_cpu_mem_usage=True,
        quantization_config=TorchAoConfig(
            FqnToConfig(
                OrderedDict(
                    [
                        (
                            r"re:.*transformer_blocks\.\d+\.ff.*",
                            Int8WeightOnlyConfig(version=2, granularity=PerTensor()),
                        ),
                        (
                            r"re:.*transformer_blocks\.\d+\.attn.*",
                            Int8WeightOnlyConfig(version=2, granularity=PerGroup(64)),
                        ),
                        ("_default", None),
                    ]
                )
            )
        ),
    )
```

### Logs

```shell
Traceback (most recent call last):
  File "/tmp/reproduce.py", line 9, in <module>
    _t = Flux2Transformer2DModel.from_pretrained(
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/repro/.venv/lib/python3.12/site-packages/huggingface_hub/utils/_validators.py", line 88, in _inner_fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/tmp/repro/.venv/lib/python3.12/site-packages/diffusers/models/modeling_utils.py", line 1398, in from_pretrained
    ) = cls._load_pretrained_model(
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/repro/.venv/lib/python3.12/site-packages/diffusers/models/modeling_utils.py", line 1792, in _load_pretrained_model
    offload_index, state_dict_index, _mismatched_keys, _error_msgs = load_fn(shard_file)
                                                                     ^^^^^^^^^^^^^^^^^^^
  File "/tmp/repro/.venv/lib/python3.12/site-packages/diffusers/models/model_loading_utils.py", line 371, in _load_shard_file
    offload_index, state_dict_index = load_model_dict_into_meta(
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/repro/.venv/lib/python3.12/site-packages/diffusers/models/model_loading_utils.py", line 303, in load_model_dict_into_meta
    hf_quantizer.create_quantized_param(
  File "/tmp/repro/.venv/lib/python3.12/site-packages/diffusers/quantizers/torchao/torchao_quantizer.py", line 366, in create_quantized_param
    quantize_(module, self.quantization_config.get_apply_tensor_subclass())
  File "/tmp/repro/.venv/lib/python3.12/site-packages/torchao/quantization/quant_api.py", line 422, in quantize_
    raise ValueError(
ValueError: Custom filter_fn and FqnToConfig were both specified. Only filter_fn=None is supported when FqnToConfig is specified.
```

### System Info

```console

Copy-and-paste the text below in your GitHub issue and FILL OUT the two last points.

- 🤗 Diffusers version: 0.40.0.dev0
- Platform: Linux-7.2.2-arch1-1-x86_64-with-glibc2.44
- Running on Google Colab?: No
- Python version: 3.12.13
- PyTorch version (GPU?): 2.13.0+rocm7.2 (True)
- Huggingface_hub version: 1.24.0
- Transformers version: 5.5.4
- Accelerate version: 1.12.0
- PEFT version: 0.18.1
- bitsandbytes version: 0.49.2
- optimum-quanto version: 0.2.7
- torchao version: 0.17.0+rocm7.2
- Safetensors version: 0.8.0
- xFormers version: not installed
- Accelerator: NA
- Using GPU in script?: optional
- Using distributed or parallel set-up in script?: no
```

### Who can help?

@sayakpaul @DN6

### PR DESCRIPTION
**Pull request by sayakpaul**
Fixes https://github.com/huggingface/diffusers/issues/14667. It surfaced additional issues which I have fixed as well.

I have run the TorchAO test suite and it's green.

### CODE REVIEW

File: src/diffusers/quantizers/torchao/torchao_quantizer.py [hunk 1]
```diff
@@ -51,31 +51,8 @@
     import torch
     import torch.nn as nn
 
-    if is_torch_version(">=", "2.5"):
```
Comment (sayakpaul) [inline src/diffusers/quantizers/torchao/torchao_quantizer.py:54]: Deadcode not needed.

File: src/diffusers/quantizers/torchao/torchao_quantizer.py [hunk 2]
```diff
@@ -145,6 +122,41 @@ def fuzzy_match_size(config_name: str) -> str | None:
     return None
 
 
+def _fqn_to_config_weight_sizes(config: "FqnToConfig") -> tuple[set[str | None], bool]:
```
Comment (sayakpaul) [inline src/diffusers/quantizers/torchao/torchao_quantizer.py:125]: Needed to determine CUDA warmup factor from `FqnConfig`.

File: src/diffusers/quantizers/torchao/torchao_quantizer.py [hunk 3]
```diff
@@ -145,6 +122,41 @@ def fuzzy_match_size(config_name: str) -> str | None:
...
+    Returns the size digits (as `fuzzy_match_size` reports them) of every config it maps to, along with whether the
+    config leaves modules unquantized -- either mapped to `None`, or unmatched with no `_default` to fall back on.
+    """
+    fqn_to_config = config.fqn_to_config
+    size_digits = {fuzzy_match_size(type(c).__name__) for c in fqn_to_config.values() if c is not None}
+    leaves_modules_unquantized = "_default" not in fqn_to_config or any(c is None for c in fqn_to_config.values())
+    return size_digits, leaves_modules_unquantized
+
+
+def _resolve_fqn_to_config(config: "FqnToConfig", module_fqn: str, param_fqn: str):
```
Comment (sayakpaul) [inline src/diffusers/quantizers/torchao/torchao_quantizer.py:138]: Needed so that we can properly resolve config for a single linear layer because that is how we create the quantized params, i.e., iterating over the modules and calling `quantize_()` over them individually.

File: src/diffusers/quantizers/torchao/torchao_quantizer.py [hunk 4]
```diff
@@ -236,18 +257,6 @@ def adjust_target_dtype(self, target_dtype: "torch.dtype") -> "torch.dtype":
...
-            return target_dtype
-
-        # We need one of the supported dtypes to be selected in order for accelerate to determine
-        # the total size of modules/parameters for auto device placement.
-        possible_device_maps = ["auto", "balanced", "balanced_low_0", "sequential"]
-        raise ValueError(
-            f"You have set `device_map` as one of {possible_device_maps} on a TorchAO quantized model but a suitable target dtype "
-            f"could not be inferred. The supported target_dtypes are: {SUPPORTED_TORCH_DTYPES_FOR_QUANTIZATION}. If you think the "
-            f"dtype you are using should be supported, please open an issue at https://github.com/huggingface/diffusers/issues."
-        )
```
Comment (sayakpaul) [inline src/diffusers/quantizers/torchao/torchao_quantizer.py:249]: Deadcode.

File: src/diffusers/quantizers/torchao/torchao_quantizer.py [hunk 5]
```diff
@@ -145,6 +122,41 @@ def fuzzy_match_size(config_name: str) -> str | None:
...
+    for fqn in (param_fqn, module_fqn):
+        if fqn in fqn_to_config:
+            return fqn_to_config[fqn]
+
+    for fqn in (param_fqn, module_fqn):
+        for pattern, pattern_config in fqn_to_config.items():
+            if pattern.startswith("re:") and re.fullmatch(pattern[3:], fqn):
+                return pattern_config
+
+    return fqn_to_config.get("_default", None)
```
Comment (Beinsezii) [inline src/diffusers/quantizers/torchao/torchao_quantizer.py:157]: since this is manually decoded instead of using torchao machinery might be worth adding UT to ensure equivalent behavior for different pattern matches?

**Review by Beinsezii — APPROVED** [review]
Initial tests seem to match what i expected from my own ai generated solution. Think gtg for now if I find any other corner cases I can open another issue.

**Comment by Beinsezii** [comment]
interestingly it seems to completely ignore quantization configs for the MiniMax H3 "vae" but I haven't investigated why yet.

**Comment by sayakpaul** [comment]
> interestingly it seems to completely ignore quantization configs for the MiniMax H3 "vae" but I haven't investigated why yet.

1. Do you have a reproducer?
2. Does `main` do it as expected?

File: src/diffusers/quantizers/torchao/torchao_quantizer.py (hunk 5 above)
Comment (sayakpaul) [reply src/diffusers/quantizers/torchao/torchao_quantizer.py:157]: Is there a torchao machinery for the case we're dealing with (we pass a single linear layer at a time for quantizing)?

**Comment by Beinsezii** [comment]
>     1. Do you have a reproducer?

not yet just something i noticed running H3, probably blacklisted in modular or has everything set to keep_in_f32. when i get time i can make a repro that doesn't need 200gb of weights.

>     2. Does `main` do it as expected?

No, think this is unrelated, shouldn't block PR

**Comment by sayakpaul** [comment]
If a given VAE has no linear layers then nothing will get applied btw because TorchAO only focuses linear layers.

**Comment by Beinsezii** [comment]
> If a given VAE has no linear layers then nothing will get applied btw because TorchAO only focuses linear layers.

it's a 10gb transformer with the usual vae resnet downsampling bolted onto the end https://huggingface.co/MiniMaxAI/MiniMax-H3/tree/main/vae

**Comment by sayakpaul** [comment]
Once I have better coverage on https://github.com/huggingface/diffusers/pull/14686#issuecomment-5519178173, I can look into it.

File: src/diffusers/quantizers/torchao/torchao_quantizer.py (hunk 5 above)
Comment (Beinsezii) [reply src/diffusers/quantizers/torchao/torchao_quantizer.py:157]: No, I was thinking it might be worth comparing quantization_config= results to directly calling quantize_ on the whole module to ensure this function remaps appropriately. So users of quantization_config= can expect it to behave like quantize_()

Comment (Beinsezii) [reply src/diffusers/quantizers/torchao/torchao_quantizer.py:157]: Specifically because it's applied per-layer the RE matches might be different when used through diffusers vs someone used to regular torchao, for example.

Comment (sayakpaul) [reply src/diffusers/quantizers/torchao/torchao_quantizer.py:157]: That is a fantastic idea actually. I will do that!

Comment (Beinsezii) [reply src/diffusers/quantizers/torchao/torchao_quantizer.py:157]: basically to me it would make sense if this function behaved similar to reference

https://github.com/pytorch/ao/blob/ca7488145beeda7416ceee0db43eb6f5124b2ebc/torchao/quantization/quant_api.py#L1636

from a top-level POV

Comment (Beinsezii) [reply src/diffusers/quantizers/torchao/torchao_quantizer.py:157]: so i might test vs that^ using from_pretrained and quantize_ then just assert all tensor classes are the same

Comment (sayakpaul) [reply src/diffusers/quantizers/torchao/torchao_quantizer.py:157]: Added one test.

File: src/diffusers/quantizers/torchao/torchao_quantizer.py [hunk 6]
```diff
@@ -372,8 +381,25 @@ def create_quantized_param(
                 module.extra_repr = types.MethodType(_linear_extra_repr, module)
         else:
             # As we perform quantization here, the repr of linear layers is set by TorchAO, so we don't have to do it ourselves
-            module._parameters[tensor_name] = torch.nn.Parameter(param_value).to(device=target_device)
-            quantize_(module, self.quantization_config.get_apply_tensor_subclass())
+            module._parameters[tensor_name] = torch.nn.Parameter(param_value.to(device=target_device))
```
Comment (sayakpaul) [inline src/diffusers/quantizers/torchao/torchao_quantizer.py:384]: The path for pre-quantized checkpoints also does the same.

**Comment by sayakpaul** [comment]
@DN6 a gentle ping.

File: src/diffusers/quantizers/torchao/torchao_quantizer.py [hunk 7]
```diff
@@ -51,31 +51,8 @@
...
-    else:
-        SUPPORTED_TORCH_DTYPES_FOR_QUANTIZATION = (
-            torch.int8,
-            torch.float8_e4m3fn,
-            torch.float8_e5m2,
-        )
-
 if is_torchao_available():
-    from torchao.quantization import quantize_
+    from torchao.quantization import FqnToConfig, quantize_
```
Comment (DN6) [inline src/diffusers/quantizers/torchao/torchao_quantizer.py:55]: I think we would need to bump the minimum `torchao` version to `0.15.0`. `FqnToConfig` doesn't exist in earlier versions, so this would raise an error.

**Review by DN6 — APPROVED** [review]

File: src/diffusers/quantizers/torchao/torchao_quantizer.py (hunk 7 above)
Comment (sayakpaul) [reply src/diffusers/quantizers/torchao/torchao_quantizer.py:55]: We already do:

https://github.com/huggingface/diffusers/blob/759164b7ad116e091e9d3e222211c9aa27d835f6/src/diffusers/quantizers/quantization_config.py#L597

But I just updated `setup.py` as well.
